crates/miden-lib/src/transaction/inputs.rs

```diff
@@ -178,251 +27,408 @@ enum AccountInputsType {
+        inputs.add_input_notes(tx_inputs, tx_args)?;
+
+        // native/executor account
+        let native = PartialAccount::from(tx_inputs.account());
+        inputs.add_account(&native, AccountInputsType::Native(tx_inputs.account_seed()))?;
+
+        // foreign accounts
+        for acc in tx_args.foreign_account_inputs() {
+            inputs.add_account(acc.account(), AccountInputsType::Foreign)?;
+            inputs.add_account_witness(acc.witness())?;
+        }
+
+        // any extra user-supplied advice
+        inputs.extend(tx_args.advice_inputs());
+
+        Ok(inputs)
     }
 
-    // --- account code ----------------------------------------------------
-
-    advice_inputs.extend_map([
-        // CODE_COMMITMENT -> [[ACCOUNT_PROCEDURE_DATA]]
-        (account_code.commitment(), account_code.as_elements()),
-    ]);
-
-    // --- account vault ----------------------------------------------------
-
-    advice_inputs.extend_merkle_store(account.vault().inner_nodes());
-
-    // populate advice map with Sparse Merkle Tree leaf nodes
-    advice_inputs
-        .extend_map(account.vault().leaves().map(|leaf| (leaf.hash(), leaf.to_elements())));
-
-    // --- account state ----------------------------------------------------
-
-    let account_id_key: Digest =
-        Digest::from([account_id.suffix(), account_id.prefix().as_felt(), ZERO, ZERO]);
-
-    // if a seed was provided, extend the map; otherwise inject the state
-    match account_inputs_type {
-        AccountInputsType::Native(Some(seed)) => advice_inputs.extend_map([(
-            // ACCOUNT_ID -> ACCOUNT_SEED
-            account_id_key,
-            seed.to_vec(),
-        )]),
-        AccountInputsType::Foreign =>
-        // Note: keep in sync with the start_foreign_context kernel procedure
-        {
-            advice_inputs.extend_map([(
-                // ACCOUNT_ID -> [ID_AND_NONCE, VAULT_ROOT, STORAGE_COMMITMENT, CODE_COMMITMENT]
-                account_id_key,
-                account_header.as_elements(),
-            )])
-        },
-        AccountInputsType::Native(None) => {},
+    pub fn into_inner(self) -> AdviceInputs {
+        self.0
     }
 
-    Ok(())
-}
+    // MUTATORS
+    // --------------------------------------------------------------------------------------------
 
-fn add_account_witness_to_advice_inputs(
-    witness: &AccountWitness,
-    advice_inputs: &mut AdviceInputs,
-) -> Result<(), TransactionInputError> {
-    let account_leaf = witness.leaf();
-    let account_leaf_hash = account_leaf.hash();
-
-    // populate advice map with the account's leaf
-    advice_inputs.extend_map([(account_leaf_hash, account_leaf.to_elements())]);
-    // extend the merkle store and map with account witnesses merkle path
-    advice_inputs.extend_merkle_store(witness.inner_nodes());
-    Ok(())
-}
+    pub fn extend(&mut self, adv_inputs: &AdviceInputs) {
+        self.0.extend(adv_inputs.clone());
+    }
```

Comment:
I think `adv_inputs here could be just `AdviceInputs` and this way we'd avoid the need for cloning.